### PR TITLE
[SPARK-34697][SQL] Allow DESCRIBE FUNCTION and SHOW FUNCTIONS explain about || (string concatenation operator)

### PR DESCRIPTION
**Pull request by sarutak**
### What changes were proposed in this pull request?

This PR fixes the behavior of `SHOW FUNCTIONS` and `DESCRIBE FUNCTION` for the `||` operator.
The result of `SHOW FUNCTIONS` doesn't contains `||` and `DESCRIBE FUNCTION ||` says `Function: || not found.` even though `||` is supported. 

### Why are the changes needed?

It's a bug.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Confirmed manually with the following commands.
```
spark-sql> DESCRIBE FUNCTION ||;
Function: ||
Usage: expr1 || expr2 - Returns the concatenation of `expr1` and `expr2`.

spark-sql> SHOW FUNCTIONS;
!
!=
%

...

|
||
~
```

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/functions.scala [hunk 1]
```diff
@@ -140,6 +140,9 @@ case class DescribeFunctionCommand(
             "[WHEN expr4 THEN expr5]* [ELSE expr6] END - " +
             "When `expr1` = `expr2`, returns `expr3`; " +
             "when `expr1` = `expr4`, return `expr5`; else return `expr6`.") :: Nil
+      case "||" =>
```
Comment (c21) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/functions.scala:143]: nit: shall we update the above comment in L123 `Hard code "<>", "!=", "between", and "case" ...` as well?

**Review by c21 — APPROVED** [review]
LGTM with minor comment.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/functions.scala (hunk 1 above)
Comment (sarutak) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/functions.scala:143]: Ah, right. I'll update it..

**Review by HyukjinKwon — APPROVED** [review]
LGTM2

**Comment by HyukjinKwon** [comment]
Merged to master, branch-3.1, branch-3.0, and branch-2.4.

@viirya FYI

**Comment by HyukjinKwon** [comment]
Merged to master, branch-3.1 and branch-3.0.

It has a conflict with branch-2.4 .. I wouldn't bother backporting it as it's really just doc change .. but I will leave it to you @sarutak.

cc @viirya FYI

**Comment by viirya** [comment]
@HyukjinKwon thanks for ping. @sarutak Please let me know if you want to backport it. Thanks!

**Comment by sarutak** [comment]
@HyukjinKwon @viirya Thanks for letting me know. I don't have strong will to backport this to `branch-2.4` as it's a tiny change.
BTW I also noticed that the built-in functions document doesn't explain about `||`, `<>`, `!=`, `between` and `case`.
http://spark.apache.org/docs/latest/api/sql/index.html

Should we cover them in the document?

**Comment by HyukjinKwon** [comment]
It will be automatically documented by using https://github.com/apache/spark/pull/31800/files#diff-43724b94583d554110e6fd766dd403dd79a7e175013dc06f610db695bad4432cR146 😉

**Comment by sarutak** [comment]
> It will be automatically documented by using https://github.com/apache/spark/pull/31800/files#diff-43724b94583d554110e6fd766dd403dd79a7e175013dc06f610db695bad4432cR146 wink

Is it true? I think the built-in functions document is generated using `ExpressionDescription` annotation like [this](https://github.com/apache/spark/blob/03dd33cc982ebb3de4354274ac49da31521b8195/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala#L842).

In fact, e.g, `<>` is absent from the document.

**Comment by HyukjinKwon** [comment]
Oh, right. `||` is not registered in the function registry .. yeah. We might have to manually document then ..

**Pull request closed without merging**